### ISSUE TITLE
Servlet Requests in Async Mode Suddenly Get The Wrong Values for Servlet Path and Path Info

### ISSUE THREAD

**Issue by justy152** [eclipse/jetty.project#888]
To reproduce:

1. Create an embedded Jetty server instance with a servlet context handler.
2. Add a servlet mapped to the URL pattern **/myServletPath/*** and set **asyncSupported** to true on the corresponding **ServletHolder**.
3. In the **doGet** method of the servlet, log the value of **request.getPathInfo()** and **request.getServletPath()**
4. Then call **request.startAsync()**, still in the doGet method of that servlet
5. Still in the doGet method, pass the **AsyncContext** returned from your call to startAsync() to a **Runnable** instance and submit the Runnable to an Executor (e.g. a fixed thread pool)
6. In the run method of the Runnable, log **((HttpServletRequest)context.getRequest()).getPathInfo()** and **((HttpServletRequest)context.getRequest()).getServletPath()**
7. Finally, in the run method of that Runnable, commit a response (any OK response will do) and then call **context.complete()** to finish off the async request.

Now start the server and visit **http://localhost:{port}/myServletPath/some/path/info** in your web browser.

The first set of logs in the doGet method of the servlet will show the correct values:

getServletPath(): **/myServletPath**
getPathInfo(): **/some/path/info**

The second set of logs in the run method of the Runnable that is supposed to be handling the asynchronous workload will show incorrect values:

getServletPath(): **[empty string]**
getPathInfo(): **/myServletPath/some/path/info**

The correct values can still be found in the request's async attributes, but intuitively if you start an asynchronous request with a no-arg call to **request.startAsync()**, you would not expect these properties of the request to change.

**Comment by janbartel** [comment]
You may think it's intuitive, but unfortunately it isn't ;)  

When you are in your spawned thread, you are probably no longer dispatched to the servlet - it's timing dependent on whether your thread runs first or you exit the doGet method first.  When you are no longer dispatched to a servlet, then there are several fields on the request that no longer contain meaningful information, and the pathinfo and servletpath are 2 of them.  Moreover, if your async handling involved forwarding of the request to other contexts, then the pathinfo and servletpath of the request would change depending upon the point in time that you sampled it.

In short, there are several fields on the request that are inherently volatile and racy to to access. The async context provides their info in the attributes.

I'd love to be able to point you to the place in the servlet spec or the javadoc that explains all of that, but unfortunately there isn't one: there are oblique references to it in various jiras in the servlet spec group, and in conversations amongst the spec group members but it has not been concretized anywhere AFAIK.

**Comment by justy152** [comment]
Hi @janbartel I read the 3.1 spec again and now I understand that it doesn't explicitly say what the values should be when you're not in a servlet call. I want to use the attributes instead but the spec doesn't assert what these values should be **before** you've dispatched from an AsyncContext. There seems to be a hole in the spec there - the hole starts right after you leave your initial servlet call (to do some processing with the AsyncContext in an application thread) and ends just before you enter the call to a dispatched servlet (using a dispatch call from the AsyncContext in that application thread)

So would you agree that, during this time, the only sure fire way of obtaining this information is to save it in your initial servlet and then use the saved information during that hole? Or is there something else you can recommend?

**Comment by gregw** [comment]
@justy152 yes the only sure way to access contextPath, servletPath and pathInfo is to access them within a dispatch to a servlet.  If you want them asynchronously then you will need to save them, either in your own data structure or add them as your own attributes to the request.

**Comment by janbartel** [comment]
Closing.
